src/proto/streams/recv.rs

```diff
@@ -741,12 +741,39 @@ impl Recv {
     }
 
     /// Handle remote sending an explicit RST_STREAM.
-    pub fn recv_reset(&mut self, frame: frame::Reset, stream: &mut Stream) {
+    pub fn recv_reset(
+        &mut self,
+        frame: frame::Reset,
+        stream: &mut Stream,
+        counts: &mut Counts,
+    ) -> Result<(), Error> {
+        // Reseting a stream that the user hasn't accepted is possible,
+        // but should be done with care. These streams will continue
+        // to take up memory in the accept queue, but will no longer be
+        // counted as "concurrent" streams.
+        //
+        // So, we have a separate limit for these.
+        //
+        // See https://github.com/hyperium/hyper/issues/2877
+        if stream.is_pending_accept {
+            if counts.can_inc_num_remote_reset_streams() {
+                counts.inc_num_remote_reset_streams();
+            } else {
+                tracing::warn!(
+                    "recv_reset; remotely-reset pending-accept streams reached limit ({:?})",
+                    counts.max_remote_reset_streams(),
+                );
+                return Err(Error::library_go_away(Reason::ENHANCE_YOUR_CALM));
```

Comment:
Can you drive-by expose `Error::is_library` or something like that, so we can check for `error.is_library() && error.reason() == Some(Reason::ENHANCE_YOUR_CALM)` for logging purposes?

Reply:
Would checking `!error.is_remote() && error.is_go_away()` be fine?

Reply:
There are 3 error initiators: remote, user and library, and there is no way to check for the user one either. If you don't want to expose `is_library` I'm fine with checking `!error.is_remote() && !error.is_user() && error.is_go_away() && error.reason() == Some(Reason::ENHANCE_YOUR_CALM)`

Reply:
We _can_ add the method. I was just wondering if what exists is enough.

As an alternative (or in addition to), we could add some debug data to GOAWAY frame, maybe `too_many_resets` or something. Oh, just looking, a method doesn't exist to access that either...

Reply:
Made in a separate PR: https://github.com/hyperium/h2/pull/669